sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -193,124 +193,99 @@ case class BroadcastNestedLoopJoinExec(
-
-    val matchedBuildRows = streamRdd.mapPartitionsInternal { streamedIter =>
-      val buildRows = relation.value
-      val matched = new BitSet(buildRows.length)
-      val joinedRow = new JoinedRow
-
-      streamedIter.foreach { streamedRow =>
+      case _ => // BuildLeft
+        val matchedBroadcastRows = getMatchedBroadcastRowsBitSet(streamed.execute(), relation)
+        val buf: CompactBuffer[InternalRow] = new CompactBuffer()
         var i = 0
+        val buildRows = relation.value
         while (i < buildRows.length) {
-          if (boundCondition(joinedRow(streamedRow, buildRows(i)))) {
-            matched.set(i)
+          if (matchedBroadcastRows.get(i) == exists) {
+            buf += buildRows(i).copy()
           }
           i += 1
         }
-      }
-      Seq(matched).toIterator
+        sparkContext.makeRDD(buf)
     }
+  }
 
-    val matchedBroadcastRows = matchedBuildRows.fold(
-      new BitSet(relation.value.length)
-    )(_ | _)
-
-    joinType match {
-      case LeftSemi =>
-        assert(buildSide == BuildLeft)
-        val buf: CompactBuffer[InternalRow] = new CompactBuffer()
-        var i = 0
-        val rel = relation.value
-        while (i < rel.length) {
-          if (matchedBroadcastRows.get(i)) {
-            buf += rel(i).copy()
+  /**
+   * The implementation for ExistenceJoin
+   */
+  private def existenceJoin(relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+    buildSide match {
+      case BuildRight =>
+        streamed.execute().mapPartitionsInternal { streamedIter =>
+          val buildRows = relation.value
+          val joinedRow = new JoinedRow
+
+          if (condition.isDefined) {
+            val resultRow = new GenericInternalRow(Array[Any](null))
+            streamedIter.map { row =>
+              val result = buildRows.exists(r => boundCondition(joinedRow(row, r)))
+              resultRow.setBoolean(0, result)
+              joinedRow(row, resultRow)
+            }
+          } else {
+            val resultRow = new GenericInternalRow(Array[Any](buildRows.nonEmpty))
+            streamedIter.map { row =>
+              joinedRow(row, resultRow)
+            }
           }
-          i += 1
         }
-        return sparkContext.makeRDD(buf)
-      case _: ExistenceJoin =>
+      case _ => // BuildLeft
+        val matchedBroadcastRows = getMatchedBroadcastRowsBitSet(streamed.execute(), relation)
         val buf: CompactBuffer[InternalRow] = new CompactBuffer()
         var i = 0
-        val rel = relation.value
-        while (i < rel.length) {
+        val buildRows = relation.value
+        while (i < buildRows.length) {
           val result = new GenericInternalRow(Array[Any](matchedBroadcastRows.get(i)))
-          buf += new JoinedRow(rel(i).copy(), result)
+          buf += new JoinedRow(buildRows(i).copy(), result)
           i += 1
         }
-        return sparkContext.makeRDD(buf)
-      case LeftAnti =>
```

Comment:
Where is this handled after this PR?

Reply:
It's handled in leftExistenceJoin()'s BuildLeft cases. Note the input parameter 'exists' control the behavior between left semi and left anti join.

Reply:
Do you mean the following line with `condition.isEmpty`?
```
case BuildLeft if condition.isEmpty =>
```

Reply:
Ah, got what you meant. You meant two lines (the above and the following).
```
case _ => // BuildLeft
```

Reply:
@dongjoon-hyun - I added comments here https://github.com/apache/spark/pull/31821/files#r594083488 and here https://github.com/apache/spark/pull/31821/files#r594083822 . Sorry if it's not clear at the first place.

Reply:
@dongjoon-hyun - yeah, you are right.